File: go/examples/todo.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Task struct {
	ID        int
	Text      string
	Completed bool
}

type TodoList struct {
	tasks []Task
	nextID int
}

func NewTodoList() *TodoList {
	return &TodoList{
		tasks:  make([]Task, 0),
		nextID: 1,
	}
}

func (tl *TodoList) AddTask(text string) {
	task := Task{
		ID:        tl.nextID,
		Text:      text,
		Completed: false,
	}
	tl.tasks = append(tl.tasks, task)
	tl.nextID++
	fmt.Printf("Đã thêm: %s\n", text)
}

func (tl *TodoList) RemoveTask(id int) {
	for i, task := range tl.tasks {
		if task.ID == id {
			tl.tasks = append(tl.tasks[:i], tl.tasks[i+1:]...)
			fmt.Printf("Đã xóa: %s\n", task.Text)
			return
		}
	}
	fmt.Println("Không tìm thấy task!")
}

func (tl *TodoList) CompleteTask(id int) {
	for i := range tl.tasks {
		if tl.tasks[i].ID == id {
			tl.tasks[i].Completed = true
			fmt.Printf("Đã hoàn thành: %s\n", tl.tasks[i].Text)
			return
		}
	}
	fmt.Println("Không tìm thấy task!")
}

func (tl *TodoList) ShowTasks() {
	if len(tl.tasks) == 0 {
		fmt.Println("Danh sách trống!")
		return
	}
	
	fmt.Println("\n=== Danh sách công việc ===")
	for _, task := range tl.tasks {
		status := "○"
		if task.Completed {
			status = "✓"
		}
		fmt.Printf("%d. [%s] %s\n", task.ID, status, task.Text)
	}
	fmt.Println()
}

func main() {
	todo := NewTodoList()
	reader := bufio.NewReader(os.Stdin)
	
	for {
		fmt.Println("\n1. Thêm công việc")
		fmt.Println("2. Xóa công việc")
		fmt.Println("3. Đánh dấu hoàn thành")
		fmt.Println("4. Hiển thị danh sách")
		fmt.Println("5. Thoát")
		
		fmt.Print("\nChọn hành động (1-5): ")
		choice, _ := reader.ReadString('\n')
		choice = strings.TrimSpace(choice)
		
		switch choice {
		case "1":
			fmt.Print("Nhập công việc: ")
			text, _ := reader.ReadString('\n')
			text = strings.TrimSpace(text)
			todo.AddTask(text)
			
		case "2":
			todo.ShowTasks()
			fmt.Print("Nhập ID cần xóa: ")
			idStr, _ := reader.ReadString('\n')
			id, err := strconv.Atoi(strings.TrimSpace(idStr))
			if err != nil {
				fmt.Println("Vui lòng nhập số hợp lệ!")
				continue
			}
			todo.RemoveTask(id)
			
		case "3":
			todo.ShowTasks()
			fmt.Print("Nhập ID đã hoàn thành: ")
			idStr, _ := reader.ReadString('\n')
			id, err := strconv.Atoi(strings.TrimSpace(idStr))
			if err != nil {
				fmt.Println("Vui lòng nhập số hợp lệ!")
				continue
			}
			todo.CompleteTask(id)
			
		case "4":
			todo.ShowTasks()
			
		case "5":
			fmt.Println("Tạm biệt!")
			return
			
		default:
			fmt.Println("Lựa chọn không hợp lệ!")
		}
	}
}

File: rust/examples/todo.rs
use std::io;

struct Task {
    id: usize,
    text: String,
    completed: bool,
}

struct TodoList {
    tasks: Vec<Task>,
    next_id: usize,
}

impl TodoList {
    fn new() -> TodoList {
        TodoList {
            tasks: Vec::new(),
            next_id: 1,
        }
    }
    
    fn add_task(&mut self, text: String) {
        let task = Task {
            id: self.next_id,
            text,
            completed: false,
        };
        self.tasks.push(task);
        self.next_id += 1;
        println!("Đã thêm công việc!");
    }
    
    fn remove_task(&mut self, id: usize) {
        if let Some(pos) = self.tasks.iter().position(|t| t.id == id) {
            let task = self.tasks.remove(pos);
            println!("Đã xóa: {}", task.text);
        } else {
            println!("Không tìm thấy task với ID: {}", id);
        }
    }
    
    fn complete_task(&mut self, id: usize) {
        if let Some(task) = self.tasks.iter_mut().find(|t| t.id == id) {
            task.completed = true;
            println!("Đã hoàn thành: {}", task.text);
        } else {
            println!("Không tìm thấy task với ID: {}", id);
        }
    }
    
    fn show_tasks(&self) {
        if self.tasks.is_empty() {
            println!("Danh sách trống!");
            return;
        }
        
        println!("\n=== Danh sách công việc ===");
        for task in &self.tasks {
            let status = if task.completed { "✓" } else { "○" };
            println!("{}. [{}] {}", task.id, status, task.text);
        }
        println!();
    }
}

fn main() {
    let mut todo = TodoList::new();
    
    loop {
        println!("\n1. Thêm công việc");
        println!("2. Xóa công việc");
        println!("3. Đánh dấu hoàn thành");
        println!("4. Hiển thị danh sách");
        println!("5. Thoát");
        
        print!("\nChọn hành động (1-5): ");
        io::Write::flush(&mut io::stdout()).unwrap();
        
        let mut choice = String::new();
        io::stdin()
            .read_line(&mut choice)
            .expect("Failed to read line");
        
        match choice.trim() {
            "1" => {
                print!("Nhập công việc: ");
                io::Write::flush(&mut io::stdout()).unwrap();
                let mut text = String::new();
                io::stdin()
                    .read_line(&mut text)
                    .expect("Failed to read line");
                todo.add_task(text.trim().to_string());
            }
            "2" => {
                todo.show_tasks();
                print!("Nhập ID cần xóa: ");
                io::Write::flush(&mut io::stdout()).unwrap();
                let mut id_str = String::new();
                io::stdin()
                    .read_line(&mut id_str)
                    .expect("Failed to read line");
                if let Ok(id) = id_str.trim().parse() {
                    todo.remove_task(id);
                } else {
                    println!("Vui lòng nhập số hợp lệ!");
                }
            }
            "3" => {
                todo.show_tasks();
                print!("Nhập ID đã hoàn thành: ");
                io::Write::flush(&mut io::stdout()).unwrap();
                let mut id_str = String::new();
                io::stdin()
                    .read_line(&mut id_str)
                    .expect("Failed to read line");
                if let Ok(id) = id_str.trim().parse() {
                    todo.complete_task(id);
                } else {
                    println!("Vui lòng nhập số hợp lệ!");
                }
            }
            "4" => {
                todo.show_tasks();
            }
            "5" => {
                println!("Tạm biệt!");
                break;
            }
            _ => {
                println!("Lựa chọn không hợp lệ!");
            }
        }
    }
}

